openghg/store/base/_base.py

```diff
@@ -111,6 +111,150 @@ def to_data(self) -> dict:
+
+        attibutes = dataset.attrs
+
+        # Build metadata (merging fn_input_parameters with dataset.attrs)
+        metadata_initial = build_metadata(attributes=attibutes, fn_input_parameters=fn_input_parameters)
+
+        gas_data = self.prepare_surface_gas_data(
+            dataset=dataset,
+            metadata=metadata_initial,
+            update_mismatch=update_mismatch,
+            site_filepath=fn_input_parameters.get("site_filepath"),
+            species_filepath=fn_input_parameters.get("species_filepath"),
+        )
+
+        # Convert gas_data dict into list of MetadataAndData objects
+        data: list[MetadataAndData] = [
+            MetadataAndData(metadata=gd["metadata"], data=gd["data"]) for gd in (gas_data or {}).values()
+        ]
+
+        # Handle chunking
+        if chunks != {}:
+            chunks = self._check_chunks_datasource(data[0], fn_input_parameters, chunks=chunks)
+
+        self._validate_datasources(data, fn_input_parameters, filepath=None)
+
+        if chunks:
+            logger.info(f"Rechunking dataset with chunks={chunks}")
+            for datasource in data:
+                datasource.data = datasource.data.chunk(chunks)
+
+        # Align attributes vs metadata
+        self.align_metadata_attributes(data=data, update_mismatch=update_mismatch)
+
+        # Attach any additional metadata
+        self.check_info_keys(info_metadata)
+        if info_metadata is None:
+            info_metadata = {}
+        # data = self.update_metadata(data, {}, additional_metadata=info_metadata)
+
+        # Save datasources to object store
+        datasource_uuids = self.assign_data(
+            data=data,
+            if_exists=if_exists,
+            new_version=new_version,
+            compressor=compressor,
+            filters=filters,
+        )
+
+        logger.info("Completed processing dataset (parser-free).")
+        return datasource_uuids
+
+    def read_dataset(
+        self,
+        dataset: xr.Dataset,
+        source_format: str,
+        if_exists: str = "auto",
+        save_current: str = "auto",
+        overwrite: bool = False,
+        force: bool = False,
+        compressor: Any | None = None,
+        filters: Any | None = None,
+        chunks: dict | None = None,
+        update_mismatch: str = "never",
+        info_metadata: dict | None = None,
+        **kwargs: Any,
+    ) -> list[dict]:
+        """
+        Process an in-memory xarray.Dataset, update metadata, and store in the object store.
+        """
+
+        from openghg.store.spec import check_parser
+        from openghg.util import check_if_need_new_version
+
+        #  Verifies if the provided source_format is presently handled by openghg. This step can be removed to allow users to treat adding dataset as a custom dataset.
+        source_format = check_parser(self._data_type, source_format)
+
+        #  Handle legacy overwrite / force flags consistently
+        if overwrite and if_exists == "auto":
```

Comment:
If these lines of code around checking the flags are starting to be duplicated between `read_dataset` and `read_file` (and we're not able to merge the `read_file` and `read_dataset` functions) then we should split this out into another method rather than duplicating the steps.

This goes for other sections in here if the steps are the same between `read_file` and `read_dataset` and are more than 1 (or 2 at most) lines.